### PR TITLE
Add support for YS MIDI on Android

### PR DESCRIPTION
**Pull request by juhanisimola-yousician**
This adds the option to replace MIDI implementation with different one, and removes all Bluetooth awareness from Android MIDI driver. See comments for details

### CODE REVIEW

File: modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java [hunk 1]
```diff
@@ -23,40 +23,26 @@
 package com.rmsl.juce;
 
 
-import android.bluetooth.BluetoothAdapter;
```
Comment (juhanisimola-yousician) [inline modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java:26]: I basically had to remove all Bluetooth awareness from JUCE. The JUCE implementation pairs all available Bluetooth devices without asking when active, and ignores Bluetooth keyboards that have been paired outside JUCE. This is not a great feature if the user is not the only person in range with Bluetooth MIDI keyboard or has already paired the keyboard in a different app.

File: modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java [hunk 2]
```diff
@@ -78,130 +64,11 @@
...
-        private static final String bluetoothLEMidiServiceUUID = "03B80E5A-EDE8-4B33-A751-6CE34EC4C700";
-
-        private HashSet<String> bluetoothMidiDevices = new HashSet<String> ();
-        private Context appContext = null;
-    }
-
     public static class JuceMidiInputPort extends MidiReceiver implements JuceMidiPort
     {
-        private native void handleReceive (long host, byte[] msg, int offset, int count, long timestamp);
+        public static native void handleReceive (long host, byte[] msg, int offset, int count, long timestamp);
```
Comment (juhanisimola-yousician) [inline modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java:72]: Changing this to static allows using different Java implementations with the same native code

File: modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java [hunk 3]
```diff
@@ -407,55 +274,15 @@ public boolean equals (Object obj)
     }
 
     //==============================================================================
-    public static class MidiDeviceManager extends MidiManager.DeviceCallback implements MidiManager.OnDeviceOpenedListener
+    public static class JuceMidiDeviceManager extends MidiManager.DeviceCallback implements MidiManager.OnDeviceOpenedListener, MidiDeviceManager
```
Comment (juhanisimola-yousician) [inline modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java:280]: MidiDeviceManager interface is extracted to allow overriding the Java implementation

File: modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java [hunk 4]
```diff
@@ -555,14 +351,12 @@ public MidiDeviceManager (Context contextToUse)
             manager.registerDeviceCallback (this, null);
         }
 
-        protected void finalize () throws Throwable
+        public void detach () throws Throwable
```
Comment (juhanisimola-yousician) [inline modules/juce_audio_devices/native/java/app/com/rmsl/juce/JuceMidiSupport.java:358]: Forced stop to allow switching between MIDI implementations

File: modules/juce_audio_devices/native/juce_android_Midi.cpp [hunk 5]
```diff
@@ -22,312 +22,10 @@
 
 namespace juce
 {
-
```
Comment (juhanisimola-yousician) [inline modules/juce_audio_devices/native/juce_android_Midi.cpp:25]: It was an interesting discovery, debugging to figure out why changes in the Java code don't show up in the app and finding the compiled bytecode blob

**Review by Arashrezaeiyousician — APPROVED** [review]

**Pull request merged**